packages/xstate-wallet/src/store/store-indexedDB-backend.ts

```diff
@@ -0,0 +1,128 @@
+  // private _ledgers: Record<string, string | undefined> = {};
+
+  private _db;
+
+  constructor(_) {
+    if (!indexedDB) {
+      console.error("Your browser doesn't support a stable version of IndexedDB.");
+    }
+    const request = indexedDB.open('xstateWallet', 1);
+    request.onerror = event => console.error('error: ', event);
+
+    request.onupgradeneeded = event => {
+      const db = (event.target as any).result;
+      db.createObjectStore('channels');
+      db.createObjectStore('objectives', {autoIncrement: true});
+      db.createObjectStore('nonces');
+      db.createObjectStore('privateKeys');
+      db.createObjectStore('ledgers');
+    };
+
+    request.onsuccess = event => {
+      this._db = request.result;
+      console.log('DB Loaded: ' + this._db);
+    };
+  }
+
+  private async getAll(collection: string) {
+    return new Promise<any>((resolve, reject) => {
+      const request = this._db
+        .transaction([collection], 'readonly')
+        .objectStore(collection)
+        .getAll(); // TODO: fix: not returning with keys
+      request.onerror = err => reject(err);
+      request.onsuccess = event => resolve(event.target.result);
+    });
+  }
+
+  private async get(collection: string, key: string) {
+    return new Promise<any>((resolve, reject) => {
+      const request = this._db
+        .transaction([collection], 'readonly')
+        .objectStore(collection)
+        .get(key);
+      request.onerror = err => reject(err);
+      request.onsuccess = _ => resolve(request.result);
+    });
+  }
+
+  private async add(collection, key, value) {
+    return new Promise<any>((resolve, reject) => {
+      const request = this._db
+        .transaction([collection], 'readwrite')
+        .objectStore(collection)
+        .add(value, key);
+      request.onerror = err => reject(err);
+      request.onsuccess = _ => resolve(value);
+    });
+  }
+
+  // Generic Getters
+
+  public async channels() {
+    return this.getAll('channels');
+  }
+  public async objectives() {
+    return this.getAll('objectives');
+  }
+  public async nonces() {
+    return this.getAll('nonces');
+  }
+  public async privateKeys() {
+    return this.getAll('privateKeys');
+  }
+  public async ledgers() {
+    return this.getAll('ledgers');
+  }
+
+  // Individual Getters
+
+  public async getChannel(key: string) {
+    return this.get('channels', key);
+  }
+  public async getObjective(key: string) {
+    return this.get('objectives', key);
+  }
+  public async getNonce(key: string) {
+    return this.get('nonces', key);
+  }
+  public async getPrivateKey(key: string) {
+    return this.get('privateKeys', key);
+  }
+  public async getLedger(key: string) {
+    return this.get('ledgers', key);
+  }
+
+  // Individual Getters
+
+  public async setPrivateKey(key: string, value: string) {
+    return this.add('privateKeys', key, value);
+  }
+  public async setChannel(key: string, value: MemoryChannelStoreEntry) {
```

Comment:
You might need to unpack the `MemoryChannelStoreEntry` here. The `MemoryChannelStoreEntry` is a potentially large object, that stores the states received for a channel. It will probably be useful to have methods like `MemoryChannelStoreEntry.addState` on the `DBBackend` itself.

(To explain some of the context here: `MemoryChannelStoreEntry` was originally `MemoryChannelStorage`, and was just an implementation detail of the `MemoryStore`, that handled operations involving the storage related to a given channel. The name was then changed to `MemoryChannelStoreEntry`, and it became the object that was passed through to the protocols, in order to get the protocols working, providing the `ChannelStoreEntry` interface that they expect. This was more a short term convenience, than a long-term architectural decision. I don't think it really makes sense to be passing the object responsible for storage through to the protocols, and we could probably separate that out here.)